### ISSUE TITLE
Syntax synchronization may not work for syntax items with pattern starting with "\n"

### ISSUE THREAD

**Issue by obcat** [vim/vim#8103]
**Describe the bug**
Syntax synchronization may not work for syntax items with pattern starting with `\n`.

**To Reproduce**

1. Run `vim --clean`.
2. Write the following three lines to the current buffer:
    ```
    foo
    TODO
    bar
    ```
    <img width="191" alt="ss 2021-04-13 9 38 15" src="https://user-images.githubusercontent.com/64692680/114479846-4389b080-9c3c-11eb-8abc-c4c7c28913b3.png">

3. Source the following script:
    ```vim
    syntax sync fromstart
    syntax match Todo "\nTODO"
    ```
    `TODO` is highlighted:

    <img width="189" alt="ss 2021-04-13 9 38 53" src="https://user-images.githubusercontent.com/64692680/114479918-69af5080-9c3c-11eb-8a42-b4d0a2dd5aeb.png">

4. Press <kbd>CTRL-E</kbd><kbd>CTRL-E</kbd>.

    <img width="190" alt="ss 2021-04-13 9 39 21" src="https://user-images.githubusercontent.com/64692680/114480175-fa862c00-9c3c-11eb-977b-ebbfcef55421.png">

5. Press <kbd>CTRL-Y</kbd>.

    The highlight for `TODO` disappears (unexpected!):

    <img width="189" alt="ss 2021-04-13 9 39 43" src="https://user-images.githubusercontent.com/64692680/114480261-31f4d880-9c3d-11eb-83ee-a3dedde8b939.png">

**Expected behavior**
The highlight does not disappear.

**Environment**
 - Vim 8.2.2670
 - macOS Catalina (version 10.15.7)
 - iTerm2

**Comment by matveyt** [comment]
What's the reason behind matching `\nTodo` instead of `^Todo`?!

**Comment by obcat** [comment]
> What's the reason behind matching \nTodo instead of ^Todo?!

Originally, I used `\n` to highlight a specific string between two consecutive blank lines and one blank line. Like this.

  ```vim
  syntax match Todo "\n\n\nTODO\n\n"
  ```

I now know that the following pattern is better and does not cause any problems of syn-sync, though:

  ```vim
  syntax match Todo "^\n\nTODO\n$"
  ```

**Comment by brammool** [comment]
When Vim draws the screen it takes the first displayed line and search for matches with syntax patterns.  However, your pattern doesn't match in the first line, the search would need to start in the line above it. See the help for *:syn-sync-linebreaks".
However, it appears that this works for changing text, but not for redrawing.

**Comment by obcat** [comment]
> When Vim draws the screen it takes the first displayed line and search for matches with syntax patterns. However, your pattern doesn't match in the first line, the search would need to start in the line above it.

Thanks. Actually, I know that behavior so I used `syntax sync fromstart` to parse from the start line:

> 3. Source the following script:
>     ```vim
>    syntax sync fromstart
>    syntax match Todo "\nTODO"
>    ```

Is this incorrect use? Or, possibly, you missed this line? (If so, I apologize my lack of explicitness...)

> See the help for *:syn-sync-linebreaks".
> However, it appears that this works for changing text, but not for redrawing.

I can reproduce the problem even if I use `syntax sync linebreaks=999` (as expected?):
```vim
syntax sync fromstart
syntax sync linebreaks=999
syntax match Todo "\nTODO"
```

By the way, even if this is a bug, I don't necessarily think this needs to be fixed, because there is a workaround as I mentioned above. Instead, it would be better to mention in the help that the syntax synchronization may not work if pattern starts with `\n`.

**Comment by brammool** [comment]
> > When Vim draws the screen it takes the first displayed line and
> > search for matches with syntax patterns. However, your pattern
> > doesn't match in the first line, the search would need to start in
> > the line above it.
> 
> Thanks. Actually, I know that behavior so I used `syntax sync
> fromstart` to parse from the start line:
> 
> > 3. Source the following script:
> >     ```vim
> >    syntax sync fromstart
> >    syntax match Todo "\nTODO"
> >    ```
> 
> Is this incorrect use? Or, possibly, you missed this line? (If so, I
> apologize my lack of explicitness...)
> 
> > See the help for *:syn-sync-linebreaks".
> > However, it appears that this works for changing text, but not for redrawing.
> 
> I can reproduce the problem even if I use `syntax sync linebreaks=999`
> (as expected?):
> ```vim
> syntax sync fromstart
> syntax sync linebreaks=999
> syntax match Todo "\nTODO"
> ```
> 
> By the way, even if this is a bug, I don't necessarily think this
> needs to be fixed, because there is a workaround as I mentioned above.
> Instead, it would be better to mention in the help that the syntax
> synchronization may not work if pattern starts with `\n`.

This should be rare, but using a multi-line pattern would be expected to
work, thus using "linebreaks" also when redrawing should be implemented.
However, it will require some work and other bugs may be considered
more important.

-- 
From "know your smileys":
 C=}>;*{)) Drunk, devilish chef with a toupee in an updraft,
	   a mustache, and a double chin

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
I think this is possibly the same as https://github.com/vim/vim/issues/5518

**Comment by brammool** [comment]
Yes.  This one has more info (and pictures), so let's close #5518.
